### DISCUSSION TITLE
Really need some help using REST API with BudiBase

### DISCUSSION THREAD

**Discussion by developer-uhwi** [Budibase/budibase#3743] [Help] [answered]
Params are not being passed, so I am unable to do paging or anything of the sort. I have really invested time in Budibase but these roadblocks are really frustrating.
For example, I have added the parameters and using them in the query string.

<img width="685" alt="Screenshot 2021-12-09 at 9 33 32 AM" src="https://user-images.githubusercontent.com/10439242/145415769-f8d21123-6b46-414a-869f-8169d8b41cca.png">

So I added a next button to facilitate paging, by passing a starting parameter etc. And then refresh the data provider so the table can be updated.
<img width="2015" alt="Screenshot 2021-12-09 at 9 36 05 AM" src="https://user-images.githubusercontent.com/10439242/145416308-417f3d1b-d515-4f50-8ca4-2923362ccaed.png">

The rest api gets the data, but the parameters that are passed are still the defaults, not the updated ones.
See below Google chrome 
<img width="824" alt="Screenshot 2021-12-09 at 9 37 41 AM" src="https://user-images.githubusercontent.com/10439242/145416657-f3312e32-5008-4223-a1a8-a7fff94a5cc2.png">

Budibase has alot of promise, but I sincerely hope I am doing the query wrong.

Your help will be much appreciated.

**Comment by mjashanks** [comment, marked as answer]
Hi There.

I think you are misunderstanding the button actions - I appreciate that we are low on documentation at the moment :)

So, looking at your setup, it seems that you have
1. A REST API query, with parameters
2. A Data Provider connected to this query 
3. A Button (as per your screenshot) that Executes your rest query, and refreshes your dataprovider.

Your misunderstanding is in step 3. The `Execute Query` and `Refresh Data Provider` are completely unconnected. If you are expecting `Execute Query` to repopulate your data provider - it won't. You need to setup the data provider, to pass parameters into the query (see my video).

My video below shows how you can achieve it. I have used the GitHub API, to fetch repositories from different organisations, depending on a button click...

1. Each button click sets a state value, called `org` 
2. My Data Provider uses the state value `org` as a parameter to its datasource

i.e. Updating the state will cause the data provider to update automatically.

Hope this helps, let me know if I have not understood your issue :)

https://user-images.githubusercontent.com/3524181/145421210-9f765302-39b7-4ec9-99e3-248957e7b423.mp4

**Reply by mjashanks** [reply]
PS, we are currently updating our REST connector. It will have pagination support built in: https://github.com/Budibase/budibase/issues?q=is%3Aissue+is%3Aopen+%22REST+Connector+V2%22

Due by next month.

**Reply by developer-uhwi** [reply]
I was able to get the paging aspect working, I put a variable on the page url e.g. /contacts/:page and used the :page and use the next button to navigate to /contacts/{{add URL.page 1}}, so I would get /contacts/2 results. Seems to be working quite fine . 

Your video and steps really helped, feeling much better. 

Another problem you can clarify is when using REST Query API (Post, Put), the json in the RequestBody field is not being passed. So unable to update/create a record using the REST API. 

Whatever data inputted in the requestBody whether static or params based is still not processed i.e. not being passed to the REST endpoint.

Another explainer would be nice though 👍 👍  , just in case I am doing it wrong.

<img width="716" alt="Screenshot 2021-12-09 at 2 04 29 PM" src="https://user-images.githubusercontent.com/10439242/145459540-01251ade-3fa5-4b04-b5f9-e7d0e91de7e7.png">


<img width="829" alt="Screenshot 2021-12-09 at 2 00 48 PM" src="https://user-images.githubusercontent.com/10439242/145459097-c30600ab-8bdc-40f6-8657-8fceea2ffbb2.png">

**Reply by mjashanks** [reply]
I set up a `POST` request, with a body. I was using the GitHub API to create a new issue in a repository. Everything worked as expected. I also connected this up to a screen in the `Design` section, which allowed me to create issues via a Budibase form.

Your setup looks ok to me, as far as Budibase is concerned. How do you know that the Body is not being sent?

This is my setup, for what its worth:

![image](https://user-images.githubusercontent.com/3524181/145555022-7e2150ef-8b64-4532-b633-905c9b263bcd.png)

**Comment by developer-uhwi** [comment]


**Reply by developer-uhwi** [reply]
I have done the github test it works OK from budibase, also tested again on my custom works, seems it is a problem from my end, will work on it...many thanks.....feeling very relief. Full Steam ahead now with budibase, as soon i as resolve the issue from my end.

**Comment by developer-uhwi** [comment]
I have to reopen this seems to have work, but it is really not.
Please use https://jsonplaceholder.typicode.com/posts from https://jsonplaceholder.typicode.com/guide/ to test the creation of a post and you will see what am I referring to.

When passing the requestBody e.g
{
    "title": "foo",
    "body": "bar",
    "userId": 1
  }

and expecting a return of
{
  "id": 101,
    "title": "foo",
    "body": "bar",
    "userId": 1
  }

it only returns
{
  "id": 101
}

I suspect the body is being passed as a JSON stringified string, instead of a JSON object. 
Really urgent that this be resolved. I await your review.

**Reply by mjashanks** [reply]
I followed the guide and got the expected response first time.

```json
{
     "id": 101,
    "title": "foo",
    "body": "bar",
    "userId": 1
}
```

Did you add the header?

![image](https://user-images.githubusercontent.com/3524181/145711199-3b14a7eb-4921-4dac-b92a-9dd776b855bb.png)

**Reply by developer-uhwi** [reply]
@mjashanks Thanks for your help and patience... indeed the header content-type: application/json was required. Sometime you just automatically assume that this will be added, but not the case.

Thanks for all your help. This issue can be closed out.

**Reply by mjashanks** [reply]
You're welcome! :)

**Reply by marian001** [reply]
Hi,
I'm very new to Budibase, managed to get through some basics, but got stuck at calling REST API with data from a form...

I have POST query set up (method: Create) with a few bindings which are to be used within JSON body. (The query works if I replace bindings with actual values in the body.)

I have a screen with a form, where Type is Create and my REST query is selected as Schema. Fields of the form correspond to bindings of the REST query.
Under the form I have a button, which should be used to submit values of the form to the REST query and run it.  As a button action I chose Execute Query, selected my REST API as Datasource and my REST query as Query. However, I have no way to link form field values to query bindings. I see this:
![image](https://user-images.githubusercontent.com/6951694/190879577-9636b8f8-7fd1-4423-b448-7b23b96bceb7.png)

..although based on this tutorial https://docs.budibase.com/docs/rest-bindings , I should see this:
![image](https://user-images.githubusercontent.com/6951694/190879632-9b6f31f9-63d7-431e-8d9d-5ee607bcace2.png)

The tutorial is using data provider instead of a button, but the current situation is the same also for data provider, i.e. you can only enter Default values (i.e. when provided value is null), but there's no place for linking actual values to individual bindings. Of course, I tried to use Default values fields for this by entering handlebars like this {{ Support form.Fields.Name }} although these were not among bindable values. This attempt failed as the values I entered weren't even saved.

Can you please point me in the right direction? Thanks

BTW, I'm running self-hosted instance on DigitalOcean.

UPDATE: I just found this https://github.com/Budibase/budibase/issues/7811   ..so I guess that explains what I'm experiencing..
